pkg/controller/kfservice/kfservice_controller.go

```diff
@@ -108,29 +122,58 @@ func (r *ReconcileService) Reconcile(request reconcile.Request) (reconcile.Resul
 		return reconcile.Result{}, err
 	}
 
-	// TODO(@yuzisun): Define the desired service object
-	service := &knservingv1alpha1.Service{}
-	if err := controllerutil.SetControllerReference(instance, service, r.scheme); err != nil {
+	desiredService, err := resources.CreateKnService(instance)
+	if err != nil {
+		log.Error(err, "Failed to create desired knservice", "name", desiredService.Name)
+		r.Recorder.Eventf(instance, v1.EventTypeWarning, "InternalError", err.Error())
 		return reconcile.Result{}, err
 	}
-
-	found := &knservingv1alpha1.Service{}
-	err = r.Get(context.TODO(), types.NamespacedName{Name: service.Name, Namespace: service.Namespace}, found)
-	if err != nil && errors.IsNotFound(err) {
-		log.Info("Creating service", "namespace", service.Namespace, "name", service.Name)
-		err = r.Create(context.TODO(), service)
-		return reconcile.Result{}, err
-	} else if err != nil {
+	if err := controllerutil.SetControllerReference(instance, desiredService, r.scheme); err != nil {
 		return reconcile.Result{}, err
 	}
 
-	if !reflect.DeepEqual(service.Spec, found.Spec) {
-		found.Spec = service.Spec
-		log.Info("Updating service", "namespace", service.Namespace, "name", service.Name)
-		err = r.Update(context.TODO(), found)
-		if err != nil {
-			return reconcile.Result{}, err
-		}
+	serviceReconciler := knservice.NewServiceReconcile(r.Client)
+
+	knService, err := serviceReconciler.Reconcile(context.TODO(), desiredService)
+	if err != nil {
+		log.Error(err, "Failed to reconcile knservice", "name", desiredService.Name)
+		r.Recorder.Eventf(instance, v1.EventTypeWarning, "InternalError", err.Error())
+		return reconcile.Result{}, err
+	}
+    
+	err = r.updateStatus(instance, knService)
+	if err != nil {
+		r.Recorder.Eventf(instance, v1.EventTypeWarning, "InternalError", err.Error())
 	}
 	return reconcile.Result{}, nil
 }
+
+func (r *ReconcileService) updateStatus(before *kfservingv1alpha1.KFService, knService *knservingv1alpha1.Service) error {
+	after := before.DeepCopy()
+	if knService.Status.Address != nil {
+		after.Status.URI.Internal = knService.Status.Address.Hostname
+	}
+	if before.Spec.Canary == nil ||
+		(before.Spec.Canary.TrafficPercent == 0 && before.Spec.Canary != nil) {
+		after.Status.Default.Name = knService.Status.LatestCreatedRevisionName
```

Comment:
Does this mean the Default.Name doesn't get set in the canary case?

Reply:
in canary case we want to keep the original default revision name and user can rollback in case canary version fails.

Reply:
What I mean is in the canary case we probably want to show

Default.Name = rev1
Canary.Name = rev2

If this if check doesn't pass and goes to the else, Default.Name doesn't get set. This means that if the very first kubectl apply kfsvc has both a canary and a default, it won't update them accordingly.

I just noticed another issue here where an old Canary definition can be persisted in the status if the first check passes. We probably want to reset Status.Canary and Status.Default on each call here.

Reply:
looks need to add some validation that canary should not get set while first revision is not created yet. The workflow should be that user first create default and then iterate on canary while preserve default on spec, after that user can either promote canary if it is stable or rollback to default if it has issues by removing trafficPercent(or set to 0).

Reply:
I'll add this to my admission controller CL :)